NPE in CompilationUnitProblemFinder.java:326

[testejcnpe.zip](https://github.com/eclipse-jdt/eclipse.jdt.core/files/13731653/testejcnpe.zip)
```
Java Model Exception: java.lang.NullPointerException
	at org.eclipse.jdt.internal.core.CompilationUnitProblemFinder.process(CompilationUnitProblemFinder.java:326)
	at org.eclipse.jdt.internal.core.CompilationUnitProblemFinder.process(CompilationUnitProblemFinder.java:348)
	at org.eclipse.jdt.internal.core.ReconcileWorkingCopyOperation.makeConsistent(ReconcileWorkingCopyOperation.java:190)
	at org.eclipse.jdt.internal.core.ReconcileWorkingCopyOperation.executeOperation(ReconcileWorkingCopyOperation.java:94)
	at org.eclipse.jdt.internal.core.JavaModelOperation.run(JavaModelOperation.java:742)
	at org.eclipse.jdt.internal.core.JavaModelOperation.runOperation(JavaModelOperation.java:808)
	at org.eclipse.jdt.internal.core.CompilationUnit.reconcile(CompilationUnit.java:1326)
	at org.eclipse.jdt.internal.core.CompilationUnit.reconcile(CompilationUnit.java:1285)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.reconcile(JavaReconcilingStrategy.java:139)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy$1.run(JavaReconcilingStrategy.java:94)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:47)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.reconcile(JavaReconcilingStrategy.java:91)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.reconcile(JavaReconcilingStrategy.java:158)
	at org.eclipse.jdt.internal.ui.text.CompositeReconcilingStrategy.reconcile(CompositeReconcilingStrategy.java:94)
	at org.eclipse.jdt.internal.ui.text.JavaCompositeReconcilingStrategy.reconcile(JavaCompositeReconcilingStrategy.java:107)
	at org.eclipse.jface.text.reconciler.MonoReconciler.process(MonoReconciler.java:78)
	at org.eclipse.jface.text.reconciler.AbstractReconciler$BackgroundThread.run(AbstractReconciler.java:207)
Caused by: java.lang.NullPointerException
Caused by: java.lang.NullPointerException
```
```
java.lang.NullPointerException: Cannot read field "returnType" because "srcMethod" is null
	at org.eclipse.jdt.internal.compiler.lookup.ImplicitNullAnnotationVerifier.checkNullSpecInheritance(ImplicitNullAnnotationVerifier.java:281)
	at org.eclipse.jdt.internal.compiler.lookup.MethodVerifier15.checkNullSpecInheritance(MethodVerifier15.java:433)
	at org.eclipse.jdt.internal.compiler.lookup.MethodVerifier15.checkAgainstInheritedMethods(MethodVerifier15.java:413)
	at org.eclipse.jdt.internal.compiler.lookup.MethodVerifier15.checkMethods(MethodVerifier15.java:603)
	at org.eclipse.jdt.internal.compiler.lookup.MethodVerifier.verify(MethodVerifier.java:1017)
	at org.eclipse.jdt.internal.compiler.lookup.MethodVerifier15.verify(MethodVerifier15.java:1022)
	at org.eclipse.jdt.internal.compiler.lookup.MethodVerifier.verify(MethodVerifier.java:1026)
	at org.eclipse.jdt.internal.compiler.lookup.SourceTypeBinding.verifyMethods(SourceTypeBinding.java:3534)
	at org.eclipse.jdt.internal.compiler.lookup.SourceTypeBinding.verifyMethods(SourceTypeBinding.java:3537)
	at org.eclipse.jdt.internal.compiler.lookup.CompilationUnitScope.verifyMethods(CompilationUnitScope.java:978)
	at org.eclipse.jdt.internal.compiler.Compiler.resolve(Compiler.java:1065)
	at org.eclipse.jdt.internal.compiler.Compiler.resolve(Compiler.java:1110)
	at org.eclipse.jdt.internal.core.CompilationUnitProblemFinder.process(CompilationUnitProblemFinder.java:282)
	at org.eclipse.jdt.internal.core.CompilationUnitProblemFinder.process(CompilationUnitProblemFinder.java:348)
	at org.eclipse.jdt.internal.core.ReconcileWorkingCopyOperation.makeConsistent(ReconcileWorkingCopyOperation.java:190)
	at org.eclipse.jdt.internal.core.ReconcileWorkingCopyOperation.executeOperation(ReconcileWorkingCopyOperation.java:94)
	at org.eclipse.jdt.internal.core.JavaModelOperation.run(JavaModelOperation.java:742)
	at org.eclipse.jdt.internal.core.JavaModelOperation.runOperation(JavaModelOperation.java:808)
	at org.eclipse.jdt.internal.core.CompilationUnit.reconcile(CompilationUnit.java:1326)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.reconcile(JavaReconcilingStrategy.java:132)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy$1.run(JavaReconcilingStrategy.java:94)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:47)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.reconcile(JavaReconcilingStrategy.java:91)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.reconcile(JavaReconcilingStrategy.java:158)
	at org.eclipse.jdt.internal.ui.text.CompositeReconcilingStrategy.reconcile(CompositeReconcilingStrategy.java:94)
	at org.eclipse.jdt.internal.ui.text.JavaCompositeReconcilingStrategy.reconcile(JavaCompositeReconcilingStrategy.java:107)
	at org.eclipse.jface.text.reconciler.MonoReconciler.process(MonoReconciler.java:78)
	at org.eclipse.jface.text.reconciler.AbstractReconciler$BackgroundThread.run(AbstractReconciler.java:207)
```

eclipse.buildId=4.30.0.20231201-1200
java.version=17.0.8.1
java.vendor=Eclipse Adoptium
BootLoader constants: OS=linux, ARCH=x86_64, WS=gtk, NL=de_DE
Framework arguments:  -product org.eclipse.epp.package.java.product
Command-line arguments:  -os linux -ws gtk -arch x86_64 -product org.eclipse.epp.package.java.product


Steps to reproduce:

1. Create a simple maven project

`pom.xml`
```xml
<project xmlns="http://maven.apache.org/POM/4.0.0"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
	<modelVersion>4.0.0</modelVersion>
	<groupId>testejcnpe</groupId>
	<artifactId>testejcnpe</artifactId>
	<packaging>jar</packaging>
	<version>1.0-SNAPSHOT</version>
	<properties>
		<maven.compiler.release>17</maven.compiler.release>
	</properties>
	<dependencies>
		<dependency>
			<groupId>org.eclipse.jdt</groupId>
			<artifactId>org.eclipse.jdt.annotation</artifactId>
			<version>2.0.0</version>
			<scope>provided</scope>
			<optional>true</optional>
		</dependency>
	</dependencies>
</project>
```

`src/main/java/testejcnpe/Foo.java`:
```java
package testejcnpe;

public interface Foo {
	@org.eclipse.jdt.annotation.NonNullByDefault
	record Bar(Object foo) implements Foo { }

	@org.eclipse.jdt.annotation.Nullable Object foo();
}
```

2. Import the Project to eclipse

3. Enable in the Project Properties -> Java Compiler -> Errors/Warnings -> Null analysis:
  -  Enable annotation-based null analysis
  - Use default annotations for null specifications
  - Inherit null annotations
 

The NPE seems to be triggered by the fact that the `foo` method in the interface has a different nullness annotation than the implicit `foo` method in the Record class.

--- a/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/ImplicitNullAnnotationVerifier.java
+++ b/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/ImplicitNullAnnotationVerifier.java
@@ -25,6 +25,7 @@ import org.eclipse.jdt.internal.compiler.ast.Argument;
 import org.eclipse.jdt.internal.compiler.ast.MethodDeclaration;
 import org.eclipse.jdt.internal.compiler.ast.NullAnnotationMatching;
 import org.eclipse.jdt.internal.compiler.ast.NullAnnotationMatching.CheckMode;
+import org.eclipse.jdt.internal.compiler.ast.RecordComponent;
 import org.eclipse.jdt.internal.compiler.ast.TypeDeclaration;
 import org.eclipse.jdt.internal.compiler.impl.CompilerOptions;
 
@@ -277,9 +278,26 @@ public class ImplicitNullAnnotationVerifier {
 					if (inheritedNullnessBits != 0) {
 						if (hasReturnNonNullDefault) {
 							// both inheritance and default: check for conflict?
-							if (shouldComplain && inheritedNullnessBits == TagBits.AnnotationNullable)
-								scope.problemReporter().conflictingNullAnnotations(currentMethod, ((MethodDeclaration) srcMethod).returnType, inheritedMethod);
-							// 	still use the inherited bits to avoid incompatibility
+							if (shouldComplain && inheritedNullnessBits == TagBits.AnnotationNullable) {
+								ASTNode location = null;
+								if (srcMethod instanceof MethodDeclaration) {
+									location = ((MethodDeclaration) srcMethod).returnType;
+								} else if (currentMethod instanceof SyntheticMethodBinding) {
+									SyntheticMethodBinding synth = (SyntheticMethodBinding) currentMethod;
+									switch (synth.purpose) {
+										case SyntheticMethodBinding.FieldReadAccess:
+											if (synth.recordComponentBinding != null) {
+												RecordComponent sourceRecordComponent = synth.sourceRecordComponent();
+												if (sourceRecordComponent != null)
+													location = sourceRecordComponent.type;
+											}
+									}
+								}
+								if (location == null)
+									location = (ASTNode) scope.referenceContext(); // fallback just in case
+								scope.problemReporter().conflictingNullAnnotations(currentMethod, location, inheritedMethod);
+								// 	still use the inherited bits to avoid incompatibility
+							}
 						}
 						if (inheritedNonNullnessInfos != null && srcMethod != null) {
 							recordDeferredInheritedNullness(scope, ((MethodDeclaration) srcMethod).returnType,

--- a/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/SyntheticMethodBinding.java
+++ b/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/SyntheticMethodBinding.java
@@ -25,6 +25,7 @@ import org.eclipse.jdt.core.compiler.CharOperation;
 import org.eclipse.jdt.internal.compiler.ast.AbstractMethodDeclaration;
 import org.eclipse.jdt.internal.compiler.ast.FieldDeclaration;
 import org.eclipse.jdt.internal.compiler.ast.LambdaExpression;
+import org.eclipse.jdt.internal.compiler.ast.RecordComponent;
 import org.eclipse.jdt.internal.compiler.ast.ReferenceExpression;
 import org.eclipse.jdt.internal.compiler.ast.SwitchStatement;
 import org.eclipse.jdt.internal.compiler.classfmt.ClassFileConstants;
@@ -663,6 +664,13 @@ public class SyntheticMethodBinding extends MethodBinding {
 	@Override
 	public LambdaExpression sourceLambda() {
 		return this.lambda;
+	}
+
+	@Override
+	public RecordComponent sourceRecordComponent() {
+		if (this.recordComponentBinding != null)
+			return this.recordComponentBinding.sourceRecordComponent();
+		return null;
 	}
 
 	@Override

--- a/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/problem/messages.properties
+++ b/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/problem/messages.properties
@@ -820,7 +820,7 @@
 936 = Null comparison always yields false: The method {0} cannot return null
 937 = Redundant null check: The field {0} is specified as @{1}
 938 = Null comparison always yields false: The field {0} is specified as @{1}
-939 = The default ''@{0}'' conflicts with the inherited ''@{1}'' annotation in the overridden method from {2} 
+939 = The default ''@{0}'' conflicts with the inherited ''@{1}'' annotation in the overridden method from {2}
 940 = Conflict between inherited null annotations ''@{0}'' declared in {1} versus ''@{2}'' declared in {3} 
 941 = Redundant null check: The field {0} cannot be null at this location (ignoring concurrency)
 942 = Null comparison always yields false: The field {0} cannot be null at this location (ignoring concurrency)

--- a/org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/NullAnnotationTest.java
+++ b/org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/NullAnnotationTest.java
@@ -6245,12 +6245,12 @@ public void testBug388281_07() {
 		"1. ERROR in p2\\Sub.java (at line 6)\n" +
 		"	public Object m(Object arg) { // (a)+(b) conflict at arg and return\n" +
 		"	       ^^^^^^\n" +
-		"The default \'@NonNull\' conflicts with the inherited \'@Nullable\' annotation in the overridden method from Super \n" +
+		"The default \'@NonNull\' conflicts with the inherited \'@Nullable\' annotation in the overridden method from Super\n" +
 		"----------\n" +
 		"2. ERROR in p2\\Sub.java (at line 6)\n" +
 		"	public Object m(Object arg) { // (a)+(b) conflict at arg and return\n" +
 		"	                       ^^^\n" +
-		"The default \'@NonNull\' conflicts with the inherited \'@Nullable\' annotation in the overridden method from Super \n" +
+		"The default \'@NonNull\' conflicts with the inherited \'@Nullable\' annotation in the overridden method from Super\n" +
 		"----------\n" +
 		"3. ERROR in p2\\Sub.java (at line 7)\n" +
 		"	System.out.println(arg.toString()); // (1)\n" +
@@ -6303,12 +6303,12 @@ public void testBug388281_08() {
 		"1. ERROR in ctest\\Ctest.java (at line 8)\n" +
 		"	public String m2(Object a2) { // (a) conflict at return\n" +
 		"	       ^^^^^^\n" +
-		"The default \'@NonNull\' conflicts with the inherited \'@Nullable\' annotation in the overridden method from I \n" +
+		"The default \'@NonNull\' conflicts with the inherited \'@Nullable\' annotation in the overridden method from I\n" +
 		"----------\n" +
 		"2. ERROR in ctest\\Ctest.java (at line 11)\n" +
 		"	public String m1(Object o1, Object o2) { // (b) conflict at o1\n" +
 		"	                        ^^\n" +
-		"The default \'@NonNull\' conflicts with the inherited \'@Nullable\' annotation in the overridden method from II \n" +
+		"The default \'@NonNull\' conflicts with the inherited \'@Nullable\' annotation in the overridden method from II\n" +
 		"----------\n" +
 		"3. ERROR in ctest\\Ctest.java (at line 12)\n" +
 		"	System.out.println(o1.toString()); // (1) inherited @Nullable\n" +

--- a/org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/NullAnnotationTests21.java
+++ b/org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/NullAnnotationTests21.java
@@ -1035,4 +1035,64 @@ public class NullAnnotationTests21 extends AbstractNullAnnotationTest {
 		runner.classLibraries = this.LIBS;
 		runner.runConformTest();
 	}
+	public void testGH1771() {
+		Runner runner = new Runner();
+		runner.customOptions = getCompilerOptions();
+		runner.customOptions.put(CompilerOptions.OPTION_ReportNullSpecViolation, CompilerOptions.ERROR);
+		runner.customOptions.put(CompilerOptions.OPTION_InheritNullAnnotations, CompilerOptions.ENABLED);
+		runner.testFiles = new String[] {
+			"p/Foo.java",
+			"""
+			package p;
+
+			public interface Foo
+			{
+				@org.eclipse.jdt.annotation.NonNullByDefault
+				record Bar(Object foo) implements
+				          Foo
+				{
+				}
+
+				@org.eclipse.jdt.annotation.Nullable
+				Object foo();
+			}
+			"""
+		};
+		runner.classLibraries = this.LIBS;
+		runner.expectedCompilerLog = """
+			----------
+			1. ERROR in p\\Foo.java (at line 6)
+				record Bar(Object foo) implements
+				           ^^^^^^
+			The default '@NonNull' conflicts with the inherited '@Nullable' annotation in the overridden method from Foo
+			----------
+			""";
+		runner.runNegativeTest();
+	}
+	public void testGH1771_corrected() {
+		Runner runner = new Runner();
+		runner.customOptions = getCompilerOptions();
+		runner.customOptions.put(CompilerOptions.OPTION_ReportNullSpecViolation, CompilerOptions.ERROR);
+		runner.customOptions.put(CompilerOptions.OPTION_InheritNullAnnotations, CompilerOptions.ENABLED);
+		runner.testFiles = new String[] {
+			"p/Foo.java",
+			"""
+			package p;
+
+			public interface Foo
+			{
+				@org.eclipse.jdt.annotation.NonNullByDefault
+				record Bar(@org.eclipse.jdt.annotation.Nullable Object foo) implements
+				          Foo
+				{
+				}
+
+				@org.eclipse.jdt.annotation.Nullable
+				Object foo();
+			}
+			"""
+		};
+		runner.classLibraries = this.LIBS;
+		runner.runConformTest();
+	}
 }

